boards/shared/Src/App/App_SharedStateMachine.c

```diff
@@ -10,10 +16,59 @@ struct StateMachine
     struct World *      world;
 };
 
+#ifdef __arm__
+static StaticSemaphore_t state_tick_mutex_storage;
+static SemaphoreHandle_t state_tick_mutex;
+#endif
+
+/**
+ * Run the given tick function over the given state machine if the tick function
+ * is not null
+ *
+ * @param state_machine The state machine to run the tick function over
+ * @param tick_function The tick function to run over the state machine
+ */
+void App_SharedStateMachine_RunStateTickFunctionIfNotNull(
+    struct StateMachine *const state_machine,
+    void (*tick_function)(struct StateMachine *))
+{
+#ifdef __arm__
+    xSemaphoreTake(state_tick_mutex, portMAX_DELAY);
+#endif
+
+    if (tick_function == NULL)
+    {
+        // Nothing to do
+        return;
+    }
```

Comment:
Shouldn't grab the semaphore if the tick function is NULL because we'd never release it! Noticed it after testing it on the actual hardware.
```suggestion
    if (tick_function == NULL)
    {
        // Nothing to do
        return;
    }
    
#ifdef __arm__
    xSemaphoreTake(state_tick_mutex, portMAX_DELAY);
#endif
```

Reply:
Nice catch. I added mutex's via `pthread` on unix so I was able to write a unit test to reproduce this as well. Fixed!